### PR TITLE
IOTSafe workarounds

### PR DESCRIPTION
**Pull request by kareem-wolfssl**
# Description

Add a couple of workarounds for IOTSafe.

Fixes zd# 12944

# Testing

Tested on customer's setup, confirmed we are performing all IOTSafe operations successfully and connecting.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by danielinux** [comment]
@kareem-wolfssl this looks OK. I would normally be against introducing workarounds for broken IoT-SAFE implementations, but it looks like in this version they can be only activated explicitly via `IOTSAFE_NO_GETDATA` and `IOTSAFE_SIGN_NO_PADDING`.

I would suggest to test this with the compliant applets and the workarounds turned off, before merging.

File: wolfcrypt/src/port/iotsafe/iotsafe.c
Comment (dgarske) [inline wolfcrypt/src/port/iotsafe/iotsafe.c:885]: Please move brace outside #ifdef and explicitly add one for this else if. Have the braces mixed with #ifdef is problematic.

File: wolfcrypt/src/port/iotsafe/iotsafe.c
Comment (dgarske) [inline wolfcrypt/src/port/iotsafe/iotsafe.c:497]: Should this instead be always stripping any trailing `00` characters, not just 4 explicitly?

File: wolfcrypt/src/port/iotsafe/iotsafe.c
Comment (dgarske) [inline wolfcrypt/src/port/iotsafe/iotsafe.c:499]: Do not declare the `int` in the for loop. Please move to top of brace section. Some compilers do not like this.

File: wolfcrypt/src/port/iotsafe/iotsafe.c
Comment (dgarske) [inline wolfcrypt/src/port/iotsafe/iotsafe.c:885]: Can we not detect the no pad case by checking the value of `sig_hdr[1] == 0`? Is no padding the right term for this? Looks like the only difference is the size is omitted.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/port/iotsafe/iotsafe.c
Comment (kareem-wolfssl) [reply wolfcrypt/src/port/iotsafe/iotsafe.c:497]: This will strip all trailing zeros, this check is confirming whether the file ends in multiple 0 bytes so it knows the file has ended and needs to be stripped.
I wanted to guard against the case that the file has a single 00 byte that happens to be at the end of the chunk.

File: wolfcrypt/src/port/iotsafe/iotsafe.c
Comment (kareem-wolfssl) [reply wolfcrypt/src/port/iotsafe/iotsafe.c:885]: Padding may not be the best term for this, basically the expected case is a 16-bit length of 0x00xx, but on some platforms we get an 8-bit length.
I can't merge this with the other case if that's what you mean, as we need to use a different offset in the response for each case.

Comment (kareem-wolfssl) [reply wolfcrypt/src/port/iotsafe/iotsafe.c]: Renamed this to `IOTSAFE_SIG_8BIT_LENGTH`, as it's an 8-bit (vs 16-bit normally) length field for a signature response.

**Review by dgarske — APPROVED** [review]

**Review by danielinux — APPROVED** [review]

**Pull request merged**
